DASH UrlTemplate parses template if it contains more than one time ea…

hugohlln opened the pull request:
With DASH, it's possible to have multiple times each DASH identifier in a template (for example by using a proxy, see [Chaos streaming proxy](https://github.com/Eyevinn/chaos-stream-proxy)).
This commit allows to compile templates containing more than 4 identifiers (it was not possible before)

microkatz commented:
Hello @hugohlln,

Thank you for looking to contribute to ExoPlayer. Before we review your Pull Request, you'll need to sign the Contributor License Agreement (CLA). Thank you!

hugohlln commented:
Hi @microkatz,

Thanks for yout reply.

I've already sign the CLA :

![image](https://github.com/androidx/media/assets/58590906/b0abe828-7327-4229-bcef-75184ba3196e)

It's just that checks are not updating maybe ?

brignolij commented:
@hugohlln thanks for the fix, will be very helpful.

sebimag commented:
@hugohlln  Wonderful fix this will help my team and allow us to continue working with exoPlayer

microkatz commented:
Hi @hugohlln,

I really appreciate your patience. Also thank you to the other developers for adding your support to this pull request in the comments!

@hugohlln, would you mind adding some unit tests in UrlTemplateTest.java with some segments that have more than one identifier?

microkatz commented on libraries/exoplayer_dash/src/main/java/androidx/media3/exoplayer/dash/manifest/UrlTemplate.java:
```diff
@@ -51,11 +51,16 @@ public final class UrlTemplate {
+    int identifiersCount = 0;
+    String[] identifiersNames = { REPRESENTATION, NUMBER, BANDWIDTH, TIME };
+    for(String identifierName : identifiersNames){
+      identifiersCount += template.split("\\$" + identifierName + "\\$").length - 1;
```
Your regex pattern does not take into account the escape sequence $$. 

If your segment template was as follows "media/$$Bandwidth$$_c$$$Number$$$_mpd.m4s" with Bandwidth=650000 and Number=1, the solution would be "$Bandwidth$_c$1$_mpd.m4s". Your regex would mistake the $$Bandwidth$$ as an identifier.

hugohlln commented:
@microkatz 

Thanks for your reply.

I've just added tests in UrlTemplateTest and also changed my regex to match identifiers with escaped dollars.

All tests are sucessfully running

microkatz commented on libraries/exoplayer_dash/src/main/java/androidx/media3/exoplayer/dash/manifest/UrlTemplate.java:
```diff
@@ -51,11 +53,19 @@ public final class UrlTemplate {
+      Matcher matcher = pattern.matcher(template);
+      while(matcher.find())
+        identifiersCount++;
+    }
```
Still incorrect regex as it matches $$Number$$ when it should not.

The parseTemplate method is already running through the string and pattern matching. 

I would suggest altering the class to use ArrayList for urlPieces, identifiers, and identifierFormat tags just so that it can grow dynamically.

hugohlln replied:
Ok, sorry. I replaced arrays by ArrayLists. That should be ok now

microkatz commented on libraries/exoplayer_dash/src/main/java/androidx/media3/exoplayer/dash/manifest/UrlTemplate.java:
```diff
@@ -39,7 +40,7 @@ public final class UrlTemplate {
 
   private final String[] urlPieces;
-  private final int[] identifiers;
+  private final Integer[] identifiers;
```
Might as well make these fields ArrayLists as you have changed the typing in all the methods. That way you don't need to do a conversion in the constructor.

microkatz commented:
I'm going to send this for internal review now. You may see some more commits being added as I make changes in response to review feedback. Please refrain from pushing any more substantive changes as it will complicate the internal review - thanks!
